src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockPattern.java

```diff
@@ -0,0 +1,135 @@
+package mods.railcraft.world.level.block.entity.multiblock;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+
+import net.minecraft.advancements.criterion.BlockPredicate;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.server.ServerWorld;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Multiblock pattern. ONLY FOR CUBE-LIKE STRUCTURES!
+ *
+ * <p>You are facing it in the Z axis (BACK/FOURTH), while LEFT-RIGHT is X, and UP/DOWN is Y.
+ */
+public class MultiblockPattern {
+  private static final Logger logger = LogManager.getLogger();
+  private final int sizeX;
+  private final int sizeY;
+  private final int sizeZ;
+  private final int offsetX;
+  private final int offsetZ;
+  private final ArrayList<ArrayList<ArrayList<BlockPredicate>>> threeDarray;
```

Comment:
Might as well usage actual array here: `BlockPredicate[][][]`.
Rename `threeDarray` to `pattern`.